### ISSUE TITLE
[JIT] [Autocast] Autocast Pass induce warning in Jit Trace

### ISSUE THREAD

**Issue by leslie-fang-intel** [pytorch/pytorch#75866]
### 🐛 Describe the bug

We find after we enable the Autocast Pass as default https://github.com/pytorch/pytorch/pull/74178 . There are lots of calculation error warning in `Jit Trace`.

Here is the code to reproduce this error:
```
import torch
import torch.nn as nn

class AtenSoftmaxRepalce(nn.Module):
    def __init__(self, dim=-1):
        super(AtenSoftmaxRepalce, self).__init__()
        self.conv = torch.nn.Conv2d(3, 64, (3, 3), stride=(2, 2), padding=(1, 1), bias=False)
        self.softmax = torch.nn.Softmax(dim)

    def forward(self, x):
        x1 = self.conv(x)
        return self.softmax(x1)

model = AtenSoftmaxRepalce()
model.eval()

x = torch.rand(1, 3, 224, 224).to(torch.bfloat16)

with torch.no_grad():
    with torch.cpu.amp.autocast(cache_enabled=False):
        model = torch.jit.trace(model, x).eval()
```
Here is the details of warning message:
```
/home/lesliefang/pytorch_1_7_1/Autocast_Pytorch/pytorch/torch/jit/_trace.py:983: TracerWarning: Output nr 1. of the traced function does not match the corresponding output of the Python function. Detailed error:
Tensor-likes are not close!

Mismatched elements: 595329 / 802816 (74.2%)
Greatest absolute difference: 0.00012071523815393448 at index (0, 63, 43, 30) (up to 1e-05 allowed)
Greatest relative difference: 0.01244930820610687 at index (0, 27, 106, 34) (up to 1e-05 allowed)

  _check_trace(
```

If we disable the Autocast Pass (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/autocast.cpp#L22), we will not see this warning messages.

### Versions

And we are using stock pytorch with commit: 123297a8c02e0ebbf8b0ae3d3cb16d0dc2e350b4.

**Comment by leslie-fang-intel** [comment]
cc @jgong5 @CaoE

**Comment by CaoE** [comment]
The reason is similar to https://github.com/pytorch/pytorch/issues/75956: JIT Autocast operation list does not align with Imperative path Autocast. 
In this eaxmple, JIT Autocast pass will intert `aten::_autocast_to_full_precision` node for softmax to cast the BF16 input into FP32 since softmax is belong to `CastPolicy::fp32_set_opt_dtype` in `JIT Autocast operation list` shown as https://github.com/pytorch/pytorch/issues/75956.
While in imperative mode, `amp` will fall through the input of softmax that is, the input data type will not be changed, i.e., BF16.  
The calculation error warning should be caused by the input data type of softmax. The absolute difference and relative difference are close to the calculation errors between BF16 and FP32.
If add `x1.float()` for the input of softmax:
```
class AtenSoftmaxRepalce(nn.Module):
    def __init__(self, dim=-1):
        super(AtenSoftmaxRepalce, self).__init__()
        self.conv = torch.nn.Conv2d(3, 64, (3, 3), stride=(2, 2), padding=(1, 1), bias=False)
        self.softmax = torch.nn.Softmax(dim)

    def forward(self, x):
        x1 = self.conv(x)
        return self.softmax(x1.float())
```
or remove softmax:
```
class AtenSoftmaxRepalce(nn.Module):
    def __init__(self, dim=-1):
        super(AtenSoftmaxRepalce, self).__init__()
        self.conv = torch.nn.Conv2d(3, 64, (3, 3), stride=(2, 2), padding=(1, 1), bias=False)

    def forward(self, x):
        return self.conv(x)
```
There will be no calculation error warnings.

**Comment by davidberard98** [comment]
I believe the softmax issue is fixed in #76661. Please reopen if there's another issue.

Regarding the larger issues with autocast compatibility between eager and jit, we can continue tracking those issues in #75956.
